app/src/main/java/com/wut/indoornavigation/logic/graph/impl/GraphImpl.java

```diff
@@ -0,0 +1,241 @@
+
+        for (int i = 0; i < verticesCount; i++) {
+            final int infinity = 100000;
+            distance[i] = infinity;
+            previous[i] = -1;
+        }
+
+        int sIndex = vertices.indexOf(s);
+
+        distance[sIndex] = 0;
+
+        UnionFind Close = new UnionFindImpl(verticesCount);
+        Comparator<Vertex> comparator = new VertexComparator(heuristicFunction,vertices,distance,t);
+        PriorityQueue<Vertex> Open = new PriorityQueue<>(comparator);
+        Open.add(s);
+
+        while (!Open.isEmpty()) {
+            Vertex u = Open.peek();
+            Integer uIndex = vertices.indexOf(u);
+            for (Vertex iVertex : Open) {
+                Integer iIndex = vertices.indexOf(iVertex);
+
+                if (distance[iIndex] + heuristicFunction.Execute(vertices.get(iIndex), t) <= distance[uIndex] + heuristicFunction.Execute(vertices.get(uIndex), t)) {
+                    u = iVertex;
+                    uIndex = vertices.indexOf(u);
+                }
+            }
+
+            Open.remove(u);
+            Close.union(uIndex);
+            if (vertices.get(uIndex).equals(t)) {
+                break;
+            }
+
+            List<Vertex> outVertices = outVertices(u);
+            for (Vertex wVertex : outVertices) {
+                int wIndex = vertices.indexOf(wVertex);
+                double uwWeight = 0;
+
+                List<Edge> uOutEdges = edges.get(vertices.get(uIndex));
+                for (Edge uOutEdge : uOutEdges) {
+                    if(uOutEdge.getTo() == wVertex){
+                        uwWeight = uOutEdge.getWeight();
+                        break;
+                    }
+                }
+                if(!Close.connected(wIndex)){
+                    if(!Open.contains(wVertex)){
+                        Open.add(wVertex);
+                        distance[wIndex] = 100000;
+                    }
+                    if (distance[wIndex] > distance[uIndex] + uwWeight){
+                        distance[wIndex] = distance[uIndex] + uwWeight;
+                        previous[wIndex] = uIndex;
+                    }
+                }
+            }
+        }
+
+        int i = t.getId();
+
+        return reproducePath(previous, i);
+    }
+
+    private List<Vertex> reproducePath(int[] previous, int targetIndex) {
+        List<Vertex> result = new ArrayList<>();
+        while(targetIndex != -1){
+            result.add(0, vertices.get(targetIndex));
+            targetIndex = previous[targetIndex];
+        }
+
+        return result;
+    }
+
+    private Vertex findVertex(int id){
+        Vertex v = null;
+        for (Vertex vertex : vertices) {
+            if(vertex.getId() == id){
+                v = vertex;
+                break;
+            }
+        }
+
+        return v;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    @Override
+    public List<Vertex> aStar(int s, int t, HeuristicFunction heuristicFunction) {
```

Comment:
nie wiem czy zawsze bedzie referencja na referencje do tych wierzcholkow, to mozna tez samo id podac i bedzie ok; bo jak sie podaje wierzcholki to musza byc dokladne z równością referencji, tk zakodzilem, moze mozna zmienic xd